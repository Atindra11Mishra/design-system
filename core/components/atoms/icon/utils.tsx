import { FontVariationType } from './Icon';

const getOpticalSize = (size?: number) => {
  if (size && size <= 20) {
    return 20;
  } else if (size && size >= 48) {
    return 48;
  }

  return size;
};

const iconVariantMapper: Record<string, object> = {
  more_horiz: {
    weight: 600,
    opticalSize: 24,
  },
  delete: {
    fill: 0,
  },
  radio_button_checked: {
    fill: 0,
  },
  radio_button_unchecked: {
    fill: 0,
  },
  schedule: {
    fill: 0,
  },
  help_outline: {
    fill: 0,
  },
  check_circle_outline: {
    fill: 0,
  },
  highlight_of: {
    fill: 0,
  },
  credit_card: {
    fill: 0,
  },
  delete_outline: {
    fill: 0,
  },
  bookmark_border: {
    fill: 0,
  },
  work_outline: {
    fill: 0,
  },
  alarm: {
    fill: 0,
  },
  arrow_circle_up: {
    fill: 0,
  },
  copyright: {
    fill: 0,
  },
  query_builder: {
    fill: 0,
  },
  card_membership: {
    fill: 0,
  },
  arrow_circle_down: {
    fill: 0,
  },
  change_history: {
    fill: 0,
  },
  aspect_ratio: {
    fill: 0,
  },
  calendar_view_month: {
    fill: 0,
  },
  alarm_on: {
    fill: 0,
  },
  thumb_down_off_alt: {
    fill: 0,
  },
  important_devices: {
    fill: 0,
  },
  turned_in_not: {
    fill: 0,
  },
  add_circle_outline: {
    fill: 0,
  },
  speed: {
    fill: 0,
  },
  content_copy: {
    fill: 0,
  },
  timelapse: {
    fill: 0,
  },
};

export const getFontVariation = (iconName: string, variations?: FontVariationType, type?: string, size?: number) => {
  const defaultVariant = {
    fill: type === 'outlined' ? 0 : 1,
    weight: 400,
    grade: 0,
    opticalSize: getOpticalSize(size),
  };

  return { ...defaultVariant, ...iconVariantMapper[iconName], ...variations };
};
