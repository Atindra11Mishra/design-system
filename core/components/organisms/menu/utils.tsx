import React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  menuTriggerRef?: any,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>,
  listRef?: any,
  subListRef?: any,
  isSubMenuTrigger?: boolean,
  triggerRef?: any,
  menuID?: string
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef);
      break;
    case 'Enter':
      handleEnterKey(focusedOption, menuTriggerRef);
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      menuTriggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    case 'Tab':
      setOpenPopover?.(false);
      break;
    case 'ArrowRight':
      navigateSubMenu(isSubMenuTrigger, 'right', listRef, subListRef, triggerRef, menuID);
      break;
    case 'ArrowLeft':
      navigateSubMenu(isSubMenuTrigger, 'left', listRef, subListRef, triggerRef, menuID);
      break;
    default:
      break;
  }
};

const handleEnterKey = (focusedOption: Element | undefined, menuTriggerRef?: any) => {
  (focusedOption as HTMLElement)?.click();
  menuTriggerRef.current.focus();
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;
  }

  const targetOption = listItems[index];

  (targetOption as HTMLElement).focus();
  setFocusedOption && setFocusedOption(targetOption);
  targetOption?.scrollIntoView?.({ block: 'center' });
};

const navigateSubMenu = (
  isSubMenuTrigger?: boolean,
  direction?: string,
  listRef?: any,
  subListRef?: any,
  triggerRef?: any,
  menuID?: string
) => {
  console.log('listRef', listRef, 'menuIDmenuIDmenuID', menuID, document.querySelector(`[data-name="${menuID}"]`));

  const element = document.querySelector(`[data-name="${menuID}"]`);
  const menuPlacement = element?.getAttribute('data-placement');
  console.log('dddddddplacement', menuPlacement, 'menuID', menuID, 'element', element);

  if (isSubMenuTrigger) {
    if (
      (direction === 'right' && menuPlacement?.includes('right')) ||
      (direction === 'left' && menuPlacement?.includes('left'))
    ) {
      const listItems = subListRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
      (listItems[0] as HTMLElement).focus();
    }
  } else if (
    (direction === 'left' && menuPlacement?.includes('right')) ||
    (direction === 'right' && menuPlacement?.includes('left'))
  ) {
    triggerRef?.current.focus();
  }
};
