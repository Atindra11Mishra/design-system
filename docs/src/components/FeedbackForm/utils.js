const positiveChoices = [
  {
    label: 'Easy to understand',
    name: 'positiveChoices',
    value: 'Easy to understand',
  },
  {
    label: 'Solved my use-case',
    name: 'positiveChoices',
    value: 'Solved my use-case',
  },
  {
    label: 'Easy to implement',
    name: 'positiveChoices',
    value: 'Easy to implement',
  },
  {
    label: 'Other',
    name: 'positiveChoices',
    value: 'Other',
  },
];

const negativeChoices = [
  {
    label: 'Too complex to understand',
    name: 'negativeChoices',
    value: 'Too complex to understand',
  },
  {
    label: 'Missing information',
    name: 'negativeChoices',
    value: 'Missing information',
  },
  {
    label: 'Did not solve my use-case',
    name: 'negativeChoices',
    value: 'Did not solve my use-case',
  },
  {
    label: 'Issue with sample code',
    name: 'negativeChoices',
    value: 'Issue with sample code',
  },
];

export const getFeedbackOptions = (selectedOption) => {
  switch (selectedOption) {
    case 'Yes':
      return {
        heading: 'What did you like?',
        choiceListOptions: positiveChoices,
        Feedback_Type_ID: '9282ac39-f587-4ae1-ab0a-0c3d23461b81',
        Selected_Option_ID: 'c6eaad04-8f97-4c7e-89df-73d530db1ce0',
        Comment_Box_ID: '1487fdc5-c078-4461-8506-a8ffcfde32c6',
      };

    case 'No':
      return {
        heading: 'Oh, what went wrong?',
        choiceListOptions: negativeChoices,
        Feedback_Type_ID: '9282ac39-f587-4ae1-ab0a-0c3d23461b81',
        Selected_Option_ID: '70aa236c-2889-436a-a359-0a97c654cb29',
        Comment_Box_ID: '13ed6dde-a0b1-4848-8dba-f36dfee0dc7c',
      };

    default:
      return;
  }
};
