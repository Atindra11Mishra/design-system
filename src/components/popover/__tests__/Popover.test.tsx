import * as React from 'react';
import { render, fireEvent, waitFor, cleanup } from '@testing-library/react';
import { Button, Popover } from '@/index';
import { PopoverProps as Props } from '@/index.type';
import { testHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/testHelper';
class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
}

window.ResizeObserver = ResizeObserver;

const BooleanValue = [true, false];
const FunctionValue = jest.fn();
const trigger = (
  <Button appearance="basic" data-test="DesignSystem-PopoverTrigger">
    Open Popup
  </Button>
);

const popoverElement = <span>Popover</span>;

describe('Popover component snapshots', () => {
  const mapper = {
    trigger: valueHelper(trigger, { required: true }),
    placement: valueHelper('bottom', { required: true }),
    open: valueHelper(true, { required: true }),
    onToggle: valueHelper(FunctionValue, { required: true }),
    dark: valueHelper(BooleanValue, { required: true, iterate: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { asFragment } = render(<Popover {...attr}>{popoverElement}</Popover>);
      expect(asFragment()).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('Popover props tests', () => {
  it('renders children with dark: false', () => {
    const { getByTestId } = render(
      <Popover trigger={trigger} open={true}>
        <div data-test="DesignSystem-Popover--Content">{popoverElement}</div>
      </Popover>
    );

    expect(getByTestId('DesignSystem-Popover--Content')).toBeInTheDocument();
    expect(getByTestId('DesignSystem-Popover--Content')).not.toHaveClass('Mds-Popover--dark');
  });

  it('Popover component with prop: dark', () => {
    const { getByTestId } = render(
      <Popover trigger={trigger} dark={true} open={true}>
        {popoverElement}
      </Popover>
    );

    expect(getByTestId('DesignSystem-Popover')).toHaveClass('Mds-Popover--dark');
  });

  it('Popover component with customStyle', () => {
    const { getByTestId } = render(
      <Popover trigger={trigger} open={true} style={{ height: '100px', width: '100px' }}>
        {popoverElement}
      </Popover>
    );

    expect(getByTestId('DesignSystem-Popover')).toHaveStyle('height: 100px; width: 100px');
  });

  it('renders Popover on trigger click', async () => {
    const { getByTestId } = render(<Popover trigger={trigger}>{popoverElement}</Popover>);

    const popoverTrigger = getByTestId('DesignSystem-PopoverTrigger');
    fireEvent.click(popoverTrigger);
    await waitFor(() => {
      expect(getByTestId('DesignSystem-Popover')).toBeInTheDocument();
    });
  });

  it('renders Popover on trigger hover', async () => {
    const { getByTestId } = render(
      <Popover trigger={trigger} hoverable={true}>
        {popoverElement}
      </Popover>
    );

    const popoverTrigger = getByTestId('DesignSystem-PopoverTrigger');
    fireEvent.mouseEnter(popoverTrigger);

    await waitFor(() => {
      expect(getByTestId('DesignSystem-Popover')).toBeInTheDocument();
    });
  });

  it('Popover component with outsidePointerDown true', async () => {
    const { queryByTestId, getByTestId } = render(
      <>
        <div data-test="DesignSystem-OutsideClick" />
        <Popover dismissOptions={{ outsidePointerDown: true }} trigger={trigger}>
          {popoverElement}
        </Popover>
      </>
    );

    const popoverTrigger = getByTestId('DesignSystem-PopoverTrigger');
    fireEvent.click(popoverTrigger);

    const outsideClick = getByTestId('DesignSystem-OutsideClick');
    fireEvent.click(outsideClick);

    await waitFor(() => {
      expect(queryByTestId('DesignSystem-Popover')).not.toBeInTheDocument();
    });
  });

  it('Popover component with outsidePointerDown false', async () => {
    const { getByTestId } = render(
      <>
        <div data-test="DesignSystem-OutsideClick" />
        <Popover dismissOptions={{ outsidePointerDown: false }} trigger={trigger}>
          {popoverElement}
        </Popover>
      </>
    );

    const popoverTrigger = getByTestId('DesignSystem-PopoverTrigger');
    fireEvent.click(popoverTrigger);

    const outsideClick = getByTestId('DesignSystem-OutsideClick');
    fireEvent.click(outsideClick);

    await waitFor(() => {
      expect(getByTestId('DesignSystem-Popover')).toBeInTheDocument();
    });
  });

  it('Popover component with escapeKey', async () => {
    const { getByTestId, queryByTestId } = render(
      <>
        <div data-test="DesignSystem-OutsideClick" />
        <Popover dismissOptions={{ escapeKey: false }} trigger={trigger}>
          {popoverElement}
        </Popover>
      </>
    );

    const popoverTrigger = getByTestId('DesignSystem-PopoverTrigger');
    fireEvent.keyDown(popoverTrigger, { keyCode: 27 });

    await waitFor(() => {
      expect(queryByTestId('DesignSystem-Popover')).toBe(null);
    });
  });

  it('Popover component with open: false', async () => {
    const { queryByTestId } = render(
      <Popover trigger={trigger} open={false} onToggle={FunctionValue}>
        {popoverElement}
      </Popover>
    );

    await waitFor(() => {
      expect(queryByTestId('DesignSystem-Popover')).not.toBeInTheDocument();
    });
  });

  it('Popover component with open and onToggle', async () => {
    const open = true;
    const { getByTestId, queryByTestId, rerender } = render(
      <Popover trigger={trigger} open={open} onToggle={FunctionValue}>
        {popoverElement}
      </Popover>
    );

    await waitFor(() => {
      expect(getByTestId('DesignSystem-Popover')).toBeInTheDocument();
    });

    const popoverTrigger = getByTestId('DesignSystem-PopoverTrigger');
    fireEvent.click(popoverTrigger);
    await waitFor(() => {
      expect(FunctionValue).toHaveBeenCalled();
    });

    rerender(
      <Popover trigger={trigger} open={!open} onToggle={FunctionValue}>
        {popoverElement}
      </Popover>
    );

    cleanup();

    await waitFor(() => {
      expect(queryByTestId('DesignSystem-Popover')).not.toBeInTheDocument();
    });
  });
});
