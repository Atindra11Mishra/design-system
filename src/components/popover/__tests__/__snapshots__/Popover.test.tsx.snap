// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover component snapshots 
  trigger: "[ReactNode]", placement: "bottom", open: true, onToggle: "[Function]", dark: false
 1`] = `
<DocumentFragment>
  <button
    aria-controls="floating-ui-4"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="Next-Button Next-Button--regular Next-Button--basic Next-Button--iconAlign-left"
    data-test="DesignSystem-PopoverTrigger"
    tabindex="0"
  >
    Open Popup
  </button>
  <style />
</DocumentFragment>
`;

exports[`Popover component snapshots 
  trigger: "[ReactNode]", placement: "bottom", open: true, onToggle: "[Function]", dark: true
 1`] = `
<DocumentFragment>
  <button
    aria-controls="floating-ui-1"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="Next-Button Next-Button--regular Next-Button--basic Next-Button--iconAlign-left"
    data-test="DesignSystem-PopoverTrigger"
    tabindex="0"
  >
    Open Popup
  </button>
  <style />
</DocumentFragment>
`;
